### DISCUSSION TITLE
Autogen Nest

### DISCUSSION THREAD

**Discussion by VAibhavSHarma900** [microsoft/autogen#4813] [Show and tell]
Can you Let me Know if we can build Agents Using a single Agent in AutoGen. If yes ,after the agents are made can they be used parallelly with the Ongoing Processes?

**Comment by ekzhu** [comment]
> Can you Let me Know if we can build Agents Using a single Agent in AutoGen

Do you mean nested agents, as agents wrapped within an agent?

> after the agents are made can they be used parallelly with the Ongoing Processes?

For v0.4 API yes.

**Reply by VAibhavSHarma900** [reply]
No i mean i want to configure one Agent to make new agents on the basis of the role described by the user or specific method(function) for any specific task and make new agent on the basis of that . Secondly, if yes how can i use those newly made agents at the exact time that are made or do i need to modify the groupChat accordingly for the synchronization of the new agents in multi agent orchestration.

Please let me know how if you get my issue , if not  i will try to explain it with different perspective of my issue.

**Reply by ekzhu** [reply]
Do you mean you want an agent to create new agents? 

And you want these agents to form a group chat or to be added to an existing group chat?

**Reply by VAibhavSHarma900** [reply]
yes , exactly , recently i have made a single agent to create new agents for me , and store those newly created agents into a json file and now i am finding a method to use them to make a big project

**Reply by ekzhu** [reply]
Sounds cool Though I believe right now AutoGen does not support natively. You might want to code the agent creation logics. There has been many existing research works on this topic. ADAS comes to mind. There is a draft work on implementing ADAS using AutoGen. #4569

**Reply by VAibhavSHarma900** [reply]
Hi , the discussion you shared , i think it is mostly about a proper automated agent system , what i was talking about it that , a single agent system could make scenario based agents for micro level tasks like, i started by making a agent who is solely responsible to print a star pattern and like these and now saving them as a python code as a normal agent with all the required needs for it and then to use those automated micro agents we can light a groupchat .
If i get anything wrong form the discussion and also i am also doing which is not possible , Please let me know 
thank you.

**Reply by ekzhu** [reply]
What you described feels like ADAS, which create new agents in the form of Python code. 

cc @yeandy do you want to chime in?

**Reply by VAibhavSHarma900** [reply]
Please let me know how can , i mold my approach to get the desired result.

**Reply by VAibhavSHarma900** [reply]
hey @ekzhu , can you please give me a brief explanation about ADAS ,then on that explanation i will try to provide the comparisons on the two approaches.

**Comment by VAibhavSHarma900** [comment]
@ekzhu and this is one of the agents made by a single agent system.





# Generated Agent File for connection_string_finder5656565656
'''
import autogen
from typing import Dict
import os

os.environ["AUTOGEN_USE_DOCKER"] = "0"


def find_connection_string(content: str) -> str:
    keyword = "ConnectionString"
    lines = content.splitlines()
    for line in lines:
        if keyword in line:
            key, value = line.split("=", 1)
            if key.strip() == keyword:
                return value.strip()
    raise ValueError("ConnectionString not found in the content.")



config_list = [{'model': 'gpt-4o', 'api_key': ', 'api_type': 'azure', 'base_url': ' 'api_version': '2023-03-15-preview'}]

llm_config = {
    "config_list": config_list
}


connection_string_finder5656565656_agent = autogen.AssistantAgent(
    name="connection_string_finder5656565656_agent",
    system_message="""Purpose: Execute the connection_string_finder5656565656 method
'''
        
        Analysis and Responsibilities:
        ### Structured Analysis

#### 1. Agent's Primary Purpose
The primary purpose of the agent is to locate and extract the value of the "ConnectionString" from a given string of content. This helps in identifying the database connection string configuration in a body of text.

#### 2. Specific Role and Responsibilities
The agent's role is to:
- Parse the provided text content.
- Search for a line that contains the keyword "ConnectionString".
- Extract and return the value associated with the "ConnectionString" if found.
- Raise an error if the "ConnectionString" is not present in the provided content.

#### 3. Required Capabilities and Limitations
##### Capabilities:
- Ability to split and read the content line-by-line.
- Capability to identify the keyword "ConnectionString" in the text.
- Skill to parse and split the line to extract the key-value pair.
- Proficiency in handling and reporting errors if the keyword is not found.

##### Limitations:
- The agent only works with plaintext content; it won't be able to parse other file formats (e.g., binary files).
- It assumes that the keyword "ConnectionString" and its value are on the same line and separated by an '=' character.
- It returns only the first occurrence of the keyword "ConnectionString".

#### 4. Success Criteria for Task Execution
- The correct and complete extraction of the value associated with "ConnectionString" from the given content.
- Proper error handling by raising a ValueError if the keyword is not found in the content.

#### 5. Expected Input Parameters and Return Values
##### Input Parameters:
- `content` (type: str): A string containing the text to be parsed.

##### Return Values:
- (type: str): The value associated with "ConnectionString" if found.
- Raises a ValueError if "ConnectionString" is not found in the content.
        
        This agent is configured to execute the defined method and handle related tasks.""",
    llm_config=llm_config,
)
'''
def get_agent():
    """Return the agent instance"""
    return connection_string_finder5656565656_agent

if __name__ == "__main__":
    
    user_proxy = autogen.UserProxyAgent(
        name="user_proxy",
        human_input_mode="TERMINATE",
        max_consecutive_auto_reply=10,
        code_execution_config={
            "work_dir": "agents",
            "use_docker": False,
            "last_n_messages": 2
        }
    )
    
    
    user_proxy.initiate_chat(connection_string_finder5656565656_agent, message="Hello! Please introduce yourself.")
'''

**Reply by ekzhu** [reply]
Thanks for the info. It would be helpful to format it by placing code blocks inside \```python and \``` blocks. 

I don't have much experience with creating autogenerated agents. Would love to have @yeandy to chim in
